### PR TITLE
client: remove expect_null and cleanup the code get_or_create()

### PR DESCRIPTION
**Pull request by lxbsz**
The expect_null is never used and it will make the code more complex.
Just remove it and simplify the related code.

Signed-off-by: Xiubo Li <xiubli@redhat.com>

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [x] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/client/Client.cc [hunk 1]
```diff
@@ -13881,16 +13848,12 @@ int Client::_unlink(Inode *dir, const char *name, const UserPerm& perm)
...
-    goto fail;
+  Dentry *de = get_or_create(dir, name);
   req->set_dentry(de);
   req->dentry_drop = CEPH_CAP_FILE_SHARED;
   req->dentry_unless = CEPH_CAP_FILE_EXCL;
 
-  res = _lookup(dir, name, 0, &otherin, perm);
+  int res = _lookup(dir, name, 0, &otherin, perm);
   if (res < 0)
     goto fail;
```
Comment (ajarr) [inline src/client/Client.cc:13858]: Here and in ll_unlink() you don't remove the 'fail' label like you do in other syscalls. Any reason why?

**Review by ajarr — COMMENTED** [review]
Any idea why 'expect_null' was even introduced? 

I'd one other question. Otherwise LGTM.

**Comment by lxbsz** [comment]
> Any idea why 'expect_null' was even introduced?
> 

I checked the commit history but didn't find any one is using this since introduced by:

```
commit 4acdb0545efdd739b3c2d0998732bd5c0f0367b2
Author: Greg Farnum <greg@inktank.com>
Date:   Fri Sep 18 15:20:29 2009 -0700

    uclient: Fixes for get_or_create
```

Maybe just introduced for interim fixing and later just removed the caller but forgot this one ?


> I'd one other question. Otherwise LGTM.

File: src/client/Client.cc (hunk 1 above)
Comment (lxbsz) [reply src/client/Client.cc:13858]: Okay, I missed this one. For the others there had the compile errors so I fixed them. Will fix it soon.

**Review by ajarr — APPROVED** [review]
I see 'fail' labels not removed in _rmdir and _rename . I missed them earlier.

**Comment by ajarr** [comment]
> > Any idea why 'expect_null' was even introduced?
> 
> I checked the commit history but didn't find any one is using this since introduced by:
> 
> ```
> commit 4acdb0545efdd739b3c2d0998732bd5c0f0367b2
> Author: Greg Farnum <greg@inktank.com>
> Date:   Fri Sep 18 15:20:29 2009 -0700
> 
>     uclient: Fixes for get_or_create
> ```
> 
> Maybe just introduced for interim fixing and later just removed the caller but forgot this one ?
> 

Thanks, Xiubo for looking this up.

> > I'd one other question. Otherwise LGTM.

**Comment by lxbsz** [comment]
> I see 'fail' labels not removed in _rmdir and _rename . I missed them earlier.

@ajarr I have removed the `fail` table from `_rmdir`, while the `_rename` couldn't because there has some other code will use it. 
Thanks!

**Review by kotreshhr — APPROVED** [review]

**Review by vshankar — APPROVED** [review]

**Comment by gregsfortytwo** [comment]
> > Any idea why 'expect_null' was even introduced?
> 
> I checked the commit history but didn't find any one is using this since introduced by:
> 
> ```
> commit 4acdb0545efdd739b3c2d0998732bd5c0f0367b2
> Author: Greg Farnum <greg@inktank.com>
> Date:   Fri Sep 18 15:20:29 2009 -0700
> 
>     uclient: Fixes for get_or_create
> ```
> 
> Maybe just introduced for interim fixing and later just removed the caller but forgot this one ?

Based on the date, that's probably from when I was updating our Hadoop bindings, and HDFS has some non-posix semantics like disallowing you to create a new file when another one already exists with that name. So I bet it's from there.

**Review by vshankar — APPROVED** [review]

**Comment by vshankar** [comment]
https://tracker.ceph.com/projects/cephfs/wiki/Master#2022-Apr-14

**Pull request merged**